resolver/src/p2p/peer/actor.rs

```diff
@@ -0,0 +1,303 @@
+                            self.fetcher.cancel(&key);
+                            self.consumer.failed(key, ()).await;
+                        }
+                    }
+                },
+
+                // Handle completed server requests
+                msg = self.serves.stream() => {
+                    let (peer, id, result) = msg;
+                    Self::handle_serve(&mut sender, peer, id, result).await;
+                },
+
+                // Handle network messages
+                msg = receiver.recv() => {
+                    let (peer, msg) = match msg {
+                        Ok(msg) => msg,
+                        Err(err) => {
+                            error!(?err, "receiver closed");
+                            return;
+                        }
+                    };
+                    let msg = match wire::PeerMsg::decode(msg) {
+                        Ok(msg) => msg,
+                        Err(err) => {
+                            warn!(?err, ?peer, "decode failed");
+                            continue;
+                        }
+                    };
+                    let id = msg.id;
+                    match msg.payload {
+                        // Peer is requesting data
+                        Some(Payload::Request(request)) => {
+                            // Parse request
+                            let Ok(key) = Key::try_from(request) else {
+                                warn!(?peer, ?id, "peer invalid request");
+                                continue;
+                            };
+                            self.handle_request(peer, id, key);
+                        },
+                        // Peer is responding to a request with a full response
+                        Some(Payload::Response(response)) => {
+                            debug!(?peer, ?id, "peer response: data");
+
+                            // Get the key associate with the response, if any
+                            let Some(key) = self.fetcher.got_response(&peer, id, true) else {
+                                continue;
+                            };
+
+                            // The peer had the data, so we can deliver it to the consumer
+                            self.consumer.deliver(key, response).await;
+                        },
+                        // Peer is responding to a request with an error
+                        None => {
+                            warn!(?peer, ?id, "peer response: error");
+
+                            // Get the key associate with the response, if any
+                            let Some(key) = self.fetcher.got_response(&peer, id, false) else {
+                                continue;
+                            };
+
+                            // The peer did not have the data, so we need to try again
+                            self.fetcher.fetch(&mut sender, key, true).await.unwrap(); // Unwrap must be safe
```

Comment:
We handle error above for `fetch()` but not here? Is the idea is that captures duplicate key submission?